File: scripts/minifab_daemon.py

# MINIFAB DAEMON

import os, time


pin_mapping = { '7':'71', '11':'78', '13':'76', '15':'74', '19':'231', '21':'232', '23':'230', '35':'198', '37':'79', '12':'77', '16':'75', '18':'73', '26':'200', '32':'201', '38':'234', '40':'72' }

def digitalWrite(pin, state):
    cmd = "sudo gpioset gpiochip0 " + pin_mapping[str(pin)] + "=" + str(state)
    os.system(cmd)

def analogRead(pin):
    return int(os.popen("sudo gpioget gpiochip0 " + pin_mapping[str(pin)]).read())



stop_button_gpio = 11
stopped = False

def start():
    pass

def loop():
    # emergency stop
    global stopped
    if analogRead(stop_button_gpio) == 0 and stopped == False:
        os.system('curl -d "script=M112" http://mfv3.local/printer/gcode/script')
        stopped = True
    else:
        stopped = False

start()
while(True):
    loop()

File: scripts/src/minifab_daemon.cpp
/**
 * @author      : stanleyarn (stanleyarn@$HOSTNAME)
 * @file        : minifab_daemon
 * @created     : Thursday May 02, 2024 11:40:20 CEST
 */

#include "gpiocb1.hpp"
#include "minifab_daemon.hpp"
#include <iostream>
#include <fstream>
#include <sstream>

using namespace std;

bool spindle_activated = true;

void start()
{
  try
  {
    ifstream log_file_in(log_file_path);
    string mode;
    getline(log_file_in, mode);
    mode = mode.substr(mode.find('=') + 1, mode.size());
    if (mode == "PRINT") gpiocb1::write(PRINT_LED_PIN, HIGH);
    else if (mode == "PENPLT") gpiocb1::write(PENPLT_LED_PIN, HIGH);
    else if (mode == "MILL") gpiocb1::write(MILL_LED_PIN, HIGH);
  }
  catch(...)
  {
    cerr << "Log file not found." << endl;
  }
}

void loop()
{

  /* SPINDLE INTERRUPT */
  if (gpiocb1::read(SPINDLE_INTERRUPT_PIN) == LOW and spindle_activated)
  {
    gpiocb1::execCommand("curl -d ""script=M112"" http://mfv3.local/printer/gcode/script"); 
    spindle_activated = false;
  }
  else
  {
    spindle_activated = true;
  }
}

int main()
{
  start();
  while (true) loop();

  return 0;
}
